components/logins/src/login.rs

```diff
@@ -84,7 +331,46 @@ impl Login {
         self.guid.as_str()
     }
 
+    // Checks whether the Login is valid, without attempting to fix any fields.
+    // Returns an error if invalid data is found, even if it could have been fixed.
     pub fn check_valid(&self) -> Result<()> {
+        self.validate_and_fixup(false)?;
+        Ok(())
+    }
+
+    // Return either the existing login, a fixed-up verion, or an error.
+    // This consumes `self` to make it easy for callers to unconditionally
+    // replace a Login with an owned fixed-up version, preventing them from
+    // using one that is invalid.
+    pub fn fixup(self) -> Result<Self> {
+        match self.maybe_fixup()? {
+            None => Ok(self),
+            Some(login) => Ok(login),
+        }
+    }
+
+    // Like `fixup()` above, but takes `self` by reference and returns
+    // an Option for the fixed-up version, allowing the caller to make
+    // more choices about what to do next.
+    pub fn maybe_fixup(&self) -> Result<Option<Self>> {
+        self.validate_and_fixup(true)
+    }
+
+    // Internal helper for doing validation and fixups.
+    fn validate_and_fixup(&self, fixup: bool) -> Result<Option<Self>> {
+        // XXX TODO: we've definitely got more validation and fixups to add here!
+
+        let mut fixed = None;
+
+        // A little helper to magic a Some(self.clone()) into existence when needed.
+        macro_rules! get_fixed_or_throw {
+           ($err:expr) => {
+               if ! fixup { throw!($err) }
```

Comment:
Nit: space between ! and fixup. And general formatting. Rustfmt doesn't really touch macro bodies,unfortunately.